libs/server/Resp/AsyncProcessor.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Garnet.common;
+using Tsavorite.core;
+
+namespace Garnet.server
+{
+    /// <summary>
+    /// Server session for RESP protocol - basic commands are in this file
+    /// </summary>
+    internal sealed partial class RespServerSession : ServerSessionBase
+    {
+        /// <summary>
+        /// Whether async mode is turned on for the session
+        /// </summary>
+        bool useAsync = false;
+
+        /// <summary>
+        /// How many async operations are started and completed
+        /// </summary>
+        long asyncStarted = 0, asyncCompleted = 0;
+
+        /// <summary>
+        /// Async waiter for async operations
+        /// </summary>
+        SingleWaiterAutoResetEvent asyncWaiter = null;
+
+        /// <summary>
+        /// Semaphore for barrier command to wait for async operations to complete
+        /// </summary>
+        SemaphoreSlim asyncDone = null;
+
+
+        /// <summary>
+        /// Handle a async network GET command that goes pending
+        /// </summary>
+        /// <typeparam name="TGarnetApi"></typeparam>
+        /// <param name="storageApi"></param>
+        void NetworkGETPending<TGarnetApi>(ref TGarnetApi storageApi)
+            where TGarnetApi : IGarnetApi
+        {
+            unsafe
+            {
+                while (!RespWriteUtils.WriteError($"ASYNC {asyncStarted++}", ref dcurr, dend))
+                    SendAndReset();
+            }
+            if (asyncStarted == 1) // first async operation on the session, create the IO continuation processor
+            {
+                asyncWaiter = new()
+                {
+                    RunContinuationsAsynchronously = true
+                };
+                var _storageApi = storageApi;
+                _ = Task.Run(async () => await AsyncGetProcessor(_storageApi));
+            }
+            else
+            {
+                Debug.Assert(asyncWaiter != null);
+                asyncWaiter.Signal();
```

Comment:
checking: any issue here if multiple signals overlap, i.e. the processor lags behind?

Reply:
Good question. The processor lagging behind is fine - the processor's next call to Tsavorite's `CompletePendingWithOutputs` via `GET_CompletePending` will take responsibility for completing all the outstanding `(asyncCompleted - asyncStarted)` number of operations.